### PR TITLE
ModelRepo - Improved ModelLoader replacement, run txt2img img2img and image lab at same time

### PR DESCRIPTION
**Pull request by cobryan05**
# Description
ModelRepo attempts to take over the manual ModelLoader responsibilities.

It annoyed me that I couldn't be generating a batch of images while using Image-to-Image and Image Lab at the same time.
So now this exists.

ModelRepo provides a centralized repository in the code to organize and manager models.

Instead of directly loading and managing models in manually, the 'load' function for a model is registered with ModelRepo. Once this is done there is no more worrying managing the model, you just use it.

You use a model by opening a ModelManager model_context:

```
    def processLDSR(image):
        metadata = ImageMetadata.get_from_image(image)
        # Open a 'with' context to make use of the LDSR model
        with model_manager.model_context(ModelNames.LDSR) as ldsr:
            result = ldsr.superResolution(image,int(imgproc_ldsr_steps),str(imgproc_ldsr_pre_downSample),str(imgproc_ldsr_post_downSample))
        ImageMetadata.set_on_image(result, metadata)

        return result
```

That's it. Whenever you want to use a model, grab it from ModelRepo in a 'with' block just use the model. ModelRepo will return a wrapper to the model instead of the actual underlying model. This wrapper will ensure that the model is loaded and on its specified device when needed. It is essentially a model scheduler for the GPU.

This is still a bit of a WIP. The current scheduler logic is "One Model At A Time," which is essentially what it sounds like: Only one model may be on the GPU at a time. If one model is moved on, all others are moved off. It's simple, but it works. It easily implements "optimized" and "turbo-optimized" modes, and as long as you aren't running multiple different models simultaneously it is pretty performant.

Future improvements to the scheduler could involve adding some brains to the scheduler. eg, track model size, usage, etc, to intelligently evict models from the GPU. No need to evict the entire GPU to quickly run a 2MB model.

Features:
* Simple model management - Just supply a name and a load function, and modelRepo will handle the rest
* Background model loading - Models are loaded on a background thread
* Automatic device swapping - ModelRepo will move devices on and off the GPU automatically. 
* * Internally can replicate 'optimized' and 'optimized-turbo' modes
* * For an example, you could run multiple streamlit UIs, each one running txt2img, img2img, etc, with different custom models per-tab, and ModelRepo would ensure that as these ran in parallel that only one model is on the GPU at a time.
* * * Running multiple different models at the same time will work, but the initial scheduler is not optimized for this.
* * * ModelRepo will move models off the GPU, but cannot see model outputs such as Tensors. You unfortunately can't just run infinite parallel generations and never run out of VRAM, since each parallel generation will use more VRAM than just the model

TODO:
[x] CPU-only model support  (done, can now specify device for models to run on and CPU is just another device)
[x ] Streamlit - img2txt
[x] Streamlit - fix loading model dialog
[ ] Scheduler options

### CODE REVIEW

**Comment by cobryan05** [comment]
This is still WIP. It has not been optimized. It also wraps function calls that it knows nothing about. Everything I've tested appears to work, but there is certainly room for this to break things.

In my initial tests there is currently a small performance hit (single-digit %) but it tends to have a lower peak VRAM usage.

With this I am able to to run txt2img, img2img and image lab simultaneously, in multiple tabs. It's not super fast running multiple stuff at once, but it works. Plenty well enough to be generating a large batch count in txt2img while doing one-offs of Image Lab and Image-to-Image.

Also, startup time:  ModelRepo loads models on a background thread, so the webui should be available very fast since it shouldn't have to wait for the models to load. However, early on there is a call to `hasattr( model, "embedding_manager" )  ` which determines whether or not to show the Textual Inversion UI. This check for 'embedding_manager' requires blocking until the entire model is loaded. My SD model does not have 'embedding_manager' so I can't experiment with optimizing this, and I haven't spent the time to figure out where to get a model with embedding_manager. Can anyone just point me directly to how to get an 'embedding_manager'-having model? :)

**Comment by cobryan05** [comment]
@ZeroCool940711 - heads up, this is still a bit WIP but should simplify adding new models to webui. It tries to handle all of that shuffling of models off and on the GPU automatically. It also shares the model code between the Gradio and Streamlit UIs.

I've started Streamlit integration, though it's not complete. I still need img2txt (need to support a single load function returning multiple models) and textual inversion (I haven't even looked in to how to set this up).

**Comment by cobryan05** [comment]
@codedealer - Just a heads up on this. Still a bit WIP, but it's a pretty large patchset and is mostly finished so thought I'd try to get eyes on it while it's finished up.

It touches a lot of places so is a bit of merge conflict hell, so it'd be nice to get any comment/concerns addressed earlier rather than later before too many conflicts build up :)

File: scripts/util/ModelRepo/model.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        if to_device == self._on_device:
+            return
+        logger.debug(f"Moving {self}")
+        start_time = time.time()
+        # TODO: Move child models as well?
+        if to_device:
+            for model in [self._instance]:
+                if hasattr(model, 'to'):
+                    model.to('cuda')
```
Comment (cobryan05) [inline scripts/util/ModelRepo/model.py:122]: self.device, not 'cuda'

File: scripts/util/ModelRepo/model_loader.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from typing import Any
+import torch
+
+
+class ModelLoader:
+    def __init__(self, device: torch.device = torch.device("cuda"), max_depth: int = 0):
+        """Initialize a ModelLoader for use with ModelRepo
+
+        Args:
+            cpu_only (bool, optional): Run this model only on the CPU
```
Comment (cobryan05) [inline scripts/util/ModelRepo/model_loader.py:10]: outdated comment

**Comment by codedealer** [comment]
@cobryan05 hopefully this won't go to waste. There is work ongoing at separating backend logic from the frontend. It will be easier to incorporate it there rather than trying to shoehorn into two different UI systems.

Join our Discord server (in the readme) if you'd like to follow the progress. Thank you.
